### PR TITLE
Clonk feat rowcol

### PR DESCRIPTION
**Pull request by Clonkk**

### CODE REVIEW

**Comment by Clonkk** [comment]
> Ah, that's indeed a good point we didn't consider. man_facepalming Should we replace the when defined(gcDestructor) essentially by a check on Nim version less than 1.6 + defined gcDestructor?

Let's what happens

**Comment by Clonkk** [comment]
Weird CI error.

```
021-11-03T15:24:34.1745078Z /home/runner/work/Arraymancer/Arraymancer/arraymancer/tests/tensor/test_init.nim(27, 9) template/generic instantiation of `suite` from here
2021-11-03T15:24:34.1746427Z /home/runner/work/Arraymancer/Arraymancer/arraymancer/tests/tensor/test_init.nim(189, 10) template/generic instantiation of `test` from here
2021-11-03T15:24:34.1748487Z /home/runner/work/Arraymancer/Arraymancer/arraymancer/tests/tensor/test_init.nim(198, 17) Error: cannot instantiate: 'fromBuffer[int]'
2021-11-03T15:24:34.1761887Z stack trace: (most recent call last)
2021-11-03T15:24:34.1762994Z /tmp/nimblecache-946033505/nimscriptapi_4213673008.nim(187, 16)
2021-11-03T15:24:34.1763915Z /home/runner/work/Arraymancer/Arraymancer/arraymancer/arraymancer.nimble(191, 8) testTask
2021-11-03T15:24:34.1764935Z /home/runner/work/Arraymancer/Arraymancer/arraymancer/arraymancer.nimble(146, 10) test
2021-11-03T15:24:34.1765900Z /home/runner/work/Arraymancer/Arraymancer/nim/lib/system/nimscript.nim(260, 7) exec
2021-11-03T15:24:34.1767366Z [1m/home/runner/work/Arraymancer/Arraymancer/nim/lib/system/nimscript.nim(260, 7) [0m[31mError: [0munhandled exception: FAILED: nim c -o:build/tests_cpu -r tests/tests_cpu.nim [OSError][0m
2021-11-03T15:24:34.1774005Z      Error: Exception raised during nimble script execution
2021-11-03T15:24:34.1785638Z ##[error]Process completed with exit code 1.
2021-11-03T15:24:34.1871043Z Post job cleanup.
```

**Comment by Clonkk** [comment]
I restarted this PR (had a mild case of covid, couldn't really make progress onit). I reduced the scope on the PR to only Memory Layout utility and cursor bugs causing https://github.com/mratsim/Arraymancer/issues/535.

CI fails because cursor is not available on lower version. There's another bug caused by fromBuffer that i'm looking into.

Some things I noticed : 
* CI doesn't run on 1.6.0
* cursor is not available on lower version 
*I'm not sure  --mm:arc|orc not tested ?

**Comment by Clonkk** [comment]
Using Nim 1.6.0 I havea codegen error when I run ``nimble test_openblas`` (or ``nimble test``)

```
CC: tensor/test_operators_blas.nim
In file included from /usr/lib/gcc/x86_64-linux-gnu/9/include/immintrin.h:55,
                 from /usr/lib/gcc/x86_64-linux-gnu/9/include/x86intrin.h:32,
                 from /home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:11:
/home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c: In function ‘gebb_ukernel_int64_x86_AVX512_tensorZtest95operators95blas_33979’:
/usr/lib/gcc/x86_64-linux-gnu/9/include/avx512fintrin.h:339:1: error: inlining failed in call to always_inline ‘_mm512_setzero_si512’: target specific option mismatch
  339 | _mm512_setzero_si512 (void)
      | ^~~~~~~~~~~~~~~~~~~~
/home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:20330:11: note: called from here
20330 |  AB13_1 = _mm512_setzero_si512();
      |           ^~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/lib/gcc/x86_64-linux-gnu/9/include/immintrin.h:55,
                 from /usr/lib/gcc/x86_64-linux-gnu/9/include/x86intrin.h:32,
                 from /home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:11:
/usr/lib/gcc/x86_64-linux-gnu/9/include/avx512fintrin.h:339:1: error: inlining failed in call to always_inline ‘_mm512_setzero_si512’: target specific option mismatch
  339 | _mm512_setzero_si512 (void)
      | ^~~~~~~~~~~~~~~~~~~~
/home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:20327:11: note: called from here
20327 |  AB13_0 = _mm512_setzero_si512();
      |           ^~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/lib/gcc/x86_64-linux-gnu/9/include/immintrin.h:55,
                 from /usr/lib/gcc/x86_64-linux-gnu/9/include/x86intrin.h:32,
                 from /home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:11:
/usr/lib/gcc/x86_64-linux-gnu/9/include/avx512fintrin.h:339:1: error: inlining failed in call to always_inline ‘_mm512_setzero_si512’: target specific option mismatch
  339 | _mm512_setzero_si512 (void)
      | ^~~~~~~~~~~~~~~~~~~~
/home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c:20324:11: note: called from here
20324 |  AB12_1 = _mm512_setzero_si512();
      |           ^~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -fmax-errors=3.
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I/home/rcaillaud/.choosenim/toolchains/nim-1.6.0/lib -I/home/rcaillaud/Workspace/localws/arraymancer/tests -o /home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c.o /home/rcaillaud/.cache/nim/tests_cpu_d/@mtensor@stest_operators_blas.nim.c' failed with exit code: 1

```

File: src/arraymancer/laser/tensor/datatypes.nim [hunk 1]
```diff
@@ -165,43 +172,43 @@ func unsafe_raw_buf*[T: KnownSupportsCopyMem](t: var Tensor[T], aligned: static
...
+#   ## ⚠️ Warning:
+#   ##     At the moment Nim's builtin term-rewriting macros are not scoped.
+#   ##     All processing within the file this is called will be considered
+#   ##     unaligned. https://github.com/nim-lang/Nim/issues/7214#issuecomment-431567894.
+#   block:
+#     template trmUnsafeRawBuf{unsafe_raw_buf(x, aligned)}(x, aligned): auto =
+#       {.noRewrite.}: unsafe_raw_buf(x, false)
+#     template trmUnsafeRawOffset{unsafe_raw_offset(x, aligned)}(x, aligned): auto =
+#       {.noRewrite.}: unsafe_raw_offset(x, false)
+#     body
```
Comment (Vindaar) [inline src/arraymancer/laser/tensor/datatypes.nim:211]: Do we leave this like this? 

As far as I understand it's used nowhere, but then again term rewrite macros are weird. @mratsim?

**Comment by Vindaar** [comment]
Can you add a changelog entry for `toFlatSeq` and `toSeq1D`, as well as mention that we now use `{.cursor.}`?

File: src/arraymancer/laser/tensor/datatypes.nim (hunk 1 above)
Comment (Clonkk) [reply src/arraymancer/laser/tensor/datatypes.nim:211]: Well  I commented it exaclty for this purpose. To see if stuff breaks; evetnaully dead code should be removed IMO but macros are so hard to test that I wasn't confident enough to remove it (it's easier to un comment code than to go through git history)

Comment (Vindaar) [reply src/arraymancer/laser/tensor/datatypes.nim:209]: tbh I've been running with it commented out for months now locally and never had any issues.

Comment (Clonkk) [reply src/arraymancer/laser/tensor/datatypes.nim:209]: I do not mind either way, I'll let @mratsim decide if the code should stay or not :)

Comment (mratsim) [reply src/arraymancer/laser/tensor/datatypes.nim:209]: It should be used when stuff from Numpy, or Apache Arrow or Tensorflow or Pytorch is passed to Arraymancer, so when memowner is set to false.

File: src/arraymancer/laser/tensor/datatypes.nim [hunk 2]
```diff
@@ -165,43 +170,43 @@ func unsafe_raw_buf*[T: KnownSupportsCopyMem](t: var Tensor[T], aligned: static
   ## and that the data is aligned by LASER_MEM_ALIGN (default 64).
   unsafe_raw_offset_impl(0)
 
-func unsafe_raw_offset*[T: KnownSupportsCopyMem](t: Tensor[T], aligned: static bool = true): RawImmutableView[T] {.inline.} =
+func unsafe_raw_offset*[T: KnownSupportsCopyMem](t: Tensor[T], offset: int = 0, aligned: static bool = true): RawImmutableView[T] {.inline.} =
```
Comment (mratsim) [inline src/arraymancer/laser/tensor/datatypes.nim:173]: What's the use-case of this offset?

File: src/arraymancer/laser/tensor/initialization.nim [hunk 3]
```diff
@@ -243,23 +243,31 @@ proc fromBuffer*[T](rawBuffer: ptr UncheckedArray[T], shape: varargs[int]): Tens
...
-  initTensorMetadata(result, size, shape)
+  initTensorMetadata(result, size, shape, layout)
   cpuStorageFromBuffer(result.storage, rawBuffer, size)
 
-func toUnsafeView*[T: KnownSupportsCopyMem](t: Tensor[T], aligned: static bool = true): ptr UncheckedArray[T] {.inline.} =
+proc fromBuffer*[T](rawBuffer: pointer, shape: varargs[int]): Tensor[T] =
+  ## Call `fromBuffer` with layout = rowMajor
+  fromBuffer[T](rawBuffer, shape, rowMajor)
+
+func toUnsafeView*[T: KnownSupportsCopyMem](t: Tensor[T], offset: int = 0, aligned: static bool = true): ptr UncheckedArray[T] {.inline.} =
```
Comment (mratsim) [inline src/arraymancer/laser/tensor/initialization.nim:267]: Use-case of this offset?

Comment (Clonkk) [reply src/arraymancer/laser/tensor/initialization.nim:267]: Just a cleaner way to do ``unsafeAddr(toUnsafeView(t, x)[offset])``

File: src/arraymancer/laser/tensor/datatypes.nim (hunk 2 above)
Comment (Clonkk) [reply src/arraymancer/laser/tensor/datatypes.nim:173]: For consistency with ``toUnsafeView`` but this one is less practical and could be removed if you wish

File: src/arraymancer/laser/tensor/initialization.nim (hunk 3 above)
Comment (mratsim) [reply src/arraymancer/laser/tensor/initialization.nim:267]: But is there a need for this construct for external libraries? `unsafeAddr(toUnsafeView(t)[offset])`

Comment (Clonkk) [reply src/arraymancer/laser/tensor/initialization.nim:267]: It's for using Tensor as a high level buffer construct that gets passed to a C API (thus the ptr API). Each call of the C API uses a different part of the buffer to store sensor data. 

When hw operation is done, a simple reshape and you get sensor data ordered easily accessible. 

Might be a marginal use case but I still find it handy and just cleaner than calling an unnecessary unsafeAddr on top of unsafeView - but it's not strictly necessary so we can remove it if you wish.

Comment (Clonkk) [reply src/arraymancer/laser/tensor/initialization.nim:267]: @mratsim So in the end, what's the consensus ? Do you want me to remove offset from the API ?

Comment (mratsim) [reply src/arraymancer/laser/tensor/initialization.nim:267]: I think it's best to remove it for now.

Once a tensor is provided as a raw buffer, the end user takes complete ownership and can write their own glue code. I'd rather keep the raw/unsafe API as simple as possible.

**Pull request merged**